### PR TITLE
Remove limit on total number of task queue threads

### PR DESCRIPTION
**Pull request by hugomendix**
The restriction is removed from Mendix 9.9.0 onwards.
Also added a recommendation on choosing the right the number of threads.

### CODE REVIEW

File: content/refguide/task-queue.md [hunk 1]
```diff
@@ -111,9 +117,10 @@ The method `executeInBackground` takes two parameters: a context and a queue nam
 
 The period for a graceful shutdown of queues can be configured as a [custom runtime](custom-settings) setting in Studio Pro. 
 
-| Configuration option          | Example value | Explanation                                              |
-|-------------------------------|---------------|----------------------------------------------------------|
-| `TaskQueue.ShutdownGracePeriod` |          10000| Time in ms to wait for task to finish when shutting down.|
+| Configuration option                       | Example value | Explanation                                                                             |
+|--------------------------------------------|---------------|-----------------------------------------------------------------------------------------|
+| `TaskQueue.ShutdownGracePeriod`            |          10000| Time in ms to wait for task to finish when shutting down.                               |
+| `com.mendix.core.ProcessedTasksCleanupAge` |       86400000| Time in ms after which `ProcessedQueueTask` are deleted automatically (off by default). |
```
Comment (MarkvanMents) [inline content/refguide/task-queue.md:123]: What value should users provide if they want to turn this off? `0`?

File: content/refguide/task-queue.md [hunk 2]
```diff
@@ -56,12 +56,18 @@ Background execution is done in so called **Task Queues**. They can be created i
...
 
-    Task Queues have a number threads. Each of these threads can process one task at a time. That is, a queue will pick up as many concurrent tasks as it has threads. Whenever a task is finished, the next one will be picked up.
+Task Queues have a number threads. Each of these threads can process one task at a time. That is, a queue will pick up as many concurrent tasks as it has threads. Whenever a task is finished, the next one will be picked up.
     
-    In general, one or two threads should be enough, unless there is a large number of tasks or tasks take a long time and need to execute in parallel. Having many threads will put additional load on the database and should not be done if not needed.
-    
-    The total number of worker threads is limited to 40 (per cluster node). There is no hard limit on cluster nodes.
+#### 2.2.1 Thread Count Recommendation 
+
+In general, one or two threads should be enough, unless there is a large number of tasks or tasks take a long time and need to execute in parallel. Having many threads will put additional load on the database and should not be done if not needed.
```
Comment (MarkvanMents) [inline content/refguide/task-queue.md:63]:
```suggestion
{{% alert type="info" %}}
In versions of Mendix below 9.9.0, there is a limit of 40 threads per cluster node.
{{% /alert %}}

In general, one or two threads should be enough, unless there is a large number of tasks or tasks take a long time and need to execute in parallel. Having many threads will put additional load on the database and should not be done if not needed. Remember that, in a horizontally-scaled app, this is the number of threads for each *node*, not the total number of threads.
```

File: content/refguide/task-queue.md [hunk 3]
```diff
@@ -205,7 +213,6 @@ Task queues have the following limitations:
 * Microflows or Java actions that are executed in the background execute as soon as possible in the order they were created, but possibly in parallel. They are consumed in FIFO order, but then executed in parallel in case of multiple threads. There is no way to execute only a single microflow or Java action at any point in time (meaning, ensure tasks are run sequentially), unless the number of threads is set to 1 and there's only a single runtime node.
 * Microflows or Java actions that are executed in the background can *only* use the following types of parameters: Boolean, Integer/Long, Decimal, String, Date and time, Enumeration, committed Persistent Entity.
 * Background microflows or Java actions will start execution as soon as the transaction in which they are created is completed. This ensures that any data that is needed by the background microflow or Java action is committed as well. It is not possible to start a background microflow or Java action immediately, halfway during a transaction. Note that if the transaction is rolled back, the task is not executed at all.
```
Comment (MarkvanMents) [inline content/refguide/task-queue.md:215]:
```suggestion
* Background microflows or Java actions will start execution as soon as the transaction in which they are created is completed. This ensures that any data that is needed by the background microflow or Java action is committed as well. It is not possible to start a background microflow or Java action immediately, halfway during a transaction. Note that if the transaction is rolled back, the task is not executed at all.
* In versions of Mendix below 9.9.0, the total amount of parallelism per node is limited to 40. This means that at most 40 queues with parallelism 1 can be defined, or a single queue with parallelism 40, or somewhere in between, as long as the total does not exceed 40.
```

File: content/refguide/task-queue.md [hunk 4]
```diff
@@ -205,7 +213,6 @@ Task queues have the following limitations:
...
-* The total amount of parallelism per node is limited to 40. This means that at most 40 queues with parallelism 1 can be defined, or a single queue with parallelism 40, or somewhere in between, as long as the total does not exceed 40.
```
Comment (MarkvanMents) [inline content/refguide/task-queue.md:208]: Should be left in for customers using versions below 9.9.0

File: content/refguide/task-queue.md [hunk 5]
```diff
@@ -131,7 +138,8 @@ Tasks that have been processed, that is have completed or failed, are saved as o
 2. Verify that tasks have run successfully, or
 3. Debug the application in case of errors.
 
-`System.ProcessedQueueTasks` objects are never deleted. The user is free to delete them when desired.
+`System.ProcessedQueueTasks` objects are by default never deleted. The user is free to delete them when desired.
+Use the `com.mendix.core.ProcessedTasksCleanupAge` runtime setting if automatic cleanup is desired.
```
Comment (MarkvanMents) [inline content/refguide/task-queue.md:142]:
```suggestion
If you are using Mendix version 9.9.0 or above, you can use the `com.mendix.core.ProcessedTasksCleanupAge` runtime setting if automatic cleanup is desired.
```

File: content/refguide/task-queue.md (hunk 1 above)
Comment (MarkvanMents) [reply content/refguide/task-queue.md:123]:
```suggestion
| `com.mendix.core.ProcessedTasksCleanupAge` |       86400000| Time in ms after which `ProcessedQueueTask` are deleted automatically (off by default). <b/>*This setting was introduced in Mendix version 9.9.0* |   
```

**Review by MarkvanMents — COMMENTED** [review]
Hi Hugo (@hugomendix )
I've made some suggestions - mainly that we can't remove reference to the limit completely as some customers might be using versions below 9.9 (e.g. 9.6.3 MTS).
Can you review my suggestions and commit them if you are happy?
Thanks.

File: content/refguide/task-queue.md [hunk 6]
```diff
@@ -56,12 +56,18 @@ Background execution is done in so called **Task Queues**. They can be created i
 
 4. Enter the value for **Threads** for each cluster node (maximum 40).
```
[the remark this answers is not in the document]
Comment (hugomendix) [reply content/refguide/task-queue.md:57]: Yes.

File: content/refguide/task-queue.md (hunk 1 above)
Comment (hugomendix) [reply content/refguide/task-queue.md:123]: To turn it off, users should not specify this setting at all.

File: content/refguide/task-queue.md [hunk 7]
```diff
@@ -56,12 +56,18 @@ Background execution is done in so called **Task Queues**. They can be created i
...
-    The total number of worker threads is limited to 40 (per cluster node). There is no hard limit on cluster nodes.
+#### 2.2.1 Thread Count Recommendation 
+
+In general, one or two threads should be enough, unless there is a large number of tasks or tasks take a long time and need to execute in parallel. Having many threads will put additional load on the database and should not be done if not needed.
 
+When choosing the number of threads for a task queue, use the following guidelines:
+* If there are only a few tasks use a single thread.
+* If the tasks perform only calculations and perform no blocking calls, use no more threads than the available number of cores.
+* Only use more threads than the available number of cores if there are a lot of tasks and they perform blocking calls.
+* Keep in mind using more threads than the number of cores will require additional scheduling and not necessarily makes it faster.
```
[the remark this answers is not in the document]
Comment (hugomendix) [reply content/refguide/task-queue.md:73]: Can we rephrase "make queued tasks finish faster" into "improve the queued task throughput"?

File: content/refguide/task-queue.md (hunk 4 above)
Comment (hugomendix) [reply content/refguide/task-queue.md:208]: Right.

File: content/refguide/task-queue.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (MarkvanMents) [reply content/refguide/task-queue.md:73]: Done

File: content/refguide/task-queue.md (hunk 1 above)
Comment (MarkvanMents) [reply content/refguide/task-queue.md:127]: Does the suggestion look better now?

**Review by MarkvanMents — APPROVED** [review]

**Pull request merged**